fix(data-transfer): fix large transfer crashes; show transfer progress

innerdvations opened the pull request:
## What does it do?

### `@strapi/data-transfer`

- **Asset bytes on the wire:** stream chunks use **base64** payloads (`transfer-asset-chunk.ts`) instead of Buffer’s default JSON shape (`{ type: 'Buffer', data: […] }`), which explodes memory on `JSON.parse` for large files. Decoders still accept **legacy Buffer JSON** and plain base64 strings so older peers keep working.

- **WebSocket responses:** a shared **`replacerForTransferWebSocket`** (`transfer-websocket-json.ts`) stringifies Buffer / typed-array values as base64 in outgoing frames, so incidental binary in messages is less likely to serialize as huge number arrays.

- **Pull path (`pull.ts`):** builds asset batches with the chunk helper; optional **SHA-256 per asset** when checksums are negotiated; on **assets** step **start**, runs **`estimateAssetTotals`** and returns **`totals`** for progress (ignored by older clients).

- **Push / pull receivers:** shared decode path for asset stream rows; **`remote-destination`** batches outgoing chunks the same way.

- **`remote-source` (`createAssetsReadStream`):** processes each incoming batch through a **writable with `highWaterMark: 1`** so only one batch is processed at a time and backpressure/deadlocks around piping are avoided. **Stall detection** (`streamTimeout`, overridable from the CLI via config) resets on incoming chunks and after **downstream drain** so slow reads do not look like a stall; timers are **cleared** when an asset ends or the stream tears down. **One warning** if the remote still sends legacy Buffer JSON for chunks (detected from the parsed payload). Optional **checksum verification** on pull when enabled.

- **`remote-destination`:** mirrors the above for push, including **checksum negotiation** with the remote.

- **Checksum negotiation:** **`init`** can include **`checksums: true`**; the server answers whether it supports it. When both sides agree, pull/push attach **SHA-256** to each asset’s **`end`** and the receiver validates (or logs and continues if the peer does not support checksums).

- **Engine:** **`#progressTrackerChunks`** wraps each asset stream so **byte progress updates per chunk** (not only per file), keeps a single consumer, and avoids double-counted bytes/speed. **`getStageTotals('assets')`** (when provided) seeds **total bytes / count** before the assets stage so the CLI can show **fractions and ETA**.

- **Local source:** **`estimate-asset-totals.ts`** walks the same upload records as **`createAssetsStream`** but only sums sizes/counts (no open file handles). **`createAssetsStream`** can warn when DB rows point at **missing files**; those warnings go through **diagnostics** so the CLI can show them.

### `@strapi/strapi` CLI

- **`strapi transfer`:** reads optional **`server.transfer.remote.assetIdleTimeoutMs`** for **`--from`** (maps to **`streamTimeout`** on the remote source provider). **`--no-checksums`** turns off end-to-end asset checksum verification for remote source and destination (default is on).

- **Progress:** ora lines refresh on a short interval so **elapsed time** moves during slow byte progress; **readable** durations; for **assets**, **transferred vs total** size and **count / totalCount** when totals exist, plus a rough **ETA** after a short warm-up.

- **Diagnostics:** always writes **info** (and the rest) to the timestamped **log file** and logs the file’s **absolute path**; **console** shows **warn+** by default and **info** when **`--verbose`** (via **`consoleLevel`** on the file logger).

### `@strapi/logger`

- **`createOutputFileConfiguration`:** optional **`consoleLevel`** so the file can stay verbose while the console stays quiet.

### Examples

- **`examples/complex/config/server.ts`:** **`transfer.remote.enabled`** 
- **`examples/complex`** / **`examples/empty`:** **`engines.node`** bumped to **`>=20.0.0 <=24.x.x`**.

### Tests / CI

- Unit and provider tests cover chunks, engine progress, remote streams, checksum negotiation, and CLI behavior. **Remote pull/push** CLI tests and helpers live under **`tests/cli`** (see **`tests/cli/README.md`** for optional heavier media env vars).

**Backward compatibility:** asset **decoding** stays tolerant of legacy wire shapes; **checksums** are opt-in via negotiation and can be disabled with **`--no-checksums`**.

---

## Why is it needed?

Large **remote** transfers (especially **pull**) could **OOM**, look **stuck** (progress only advanced per file), or hit **backpressure/stall** bugs. Asset payloads as Buffer JSON made parsing especially expensive. This branch addresses those issues, improves **visibility** (progress, totals, warnings, logs), and adds **optional integrity checks** without breaking older Strapi versions on the other end.

---

## How to test it?

Latest experimental: ``0.0.0-experimental.11f10e472f3e424651fbe8ee40a844786e8d836f`

- **Unit:** `yarn test:unit` (or target `@strapi/data-transfer` / `@strapi/strapi` CLI tests).
- **CLI:** `yarn test:cli -d strapi --testPathPattern=data-transfer`
- **Remote e2e (slower):** `tests/cli/README.md` — **`pull-remote`**, **`push-remote`**, optional **`TRANSFER_CLI_MEDIA_*`**.
- **Manual:** `strapi transfer --from` / `--to` with large media — smooth byte progress, totals/ETA when both ends support them, **`--no-checksums`** if you need to rule out checksum overhead, and reasonable memory vs a pre-change build.

---

## Documentation needed

- **`server.transfer.remote.assetIdleTimeoutMs`:** optional **`config/server`**, milliseconds without forward progress while **`strapi transfer --from`** streams assets; applies to the **CLI machine** pulling from the remote. When to raise it (very large files, slow links, backpressure).

---

## Related issue(s) / PRs

- Fixes CMS-112
- Fixes #23485
- Fixes #25175

innerdvations committed:
enhancement: update progress per chunk

Convly commented on packages/core/data-transfer/src/engine/index.ts:
```diff
@@ -316,8 +316,34 @@ class TransferEngine<
+    },
+    perChunk?: boolean
   ) {
+    if (perChunk) {
```
I'm not a big fan of always adding more options and branches to the flow, could we try to simplify it a bit?

For instance, since the perChunk option seems to only be useful if coupled with the asset stage, could we remove the option and instead branch based on the stage value?

innerdvations replied:
I'd rather have the stage itself decide, since it needs to provide additional info anyway. And even though it's only on assets stage for now, I thought it would be better to keep it generic so that we could use it anywhere else in the future, no reason for it to be tied to assets necessarily.

Convly commented on packages/core/data-transfer/src/engine/index.ts:
```diff
@@ -916,10 +942,15 @@ class TransferEngine<
-      key: (value: IAsset) => extname(value.filename) || 'No extension',
-    });
+    // Use the perChunk tracker for assets
+    const tracker = this.#progressTracker(
```
Instead of modifying the generic method, we could also just have a dedicated method to create an assetProgressTracker? Might be easier to maintain

innerdvations replied:
Yeah, I can split it off into a different method instead of using an option

innerdvations committed:
fix: check stageprogress exists to make ts happy

innerdvations committed:
chore: split progress tracker into two methods

innerdvations committed:
test: fix lint

innerdvations committed:
enhancement: display readable times

innerdvations commented:
Oops, I broke the speed indicator for assets, and also noticed it's double-reporting the filesize. Fixing that now.

innerdvations committed:
fix: speed indication for assets

innerdvations committed:
fix: restore speed indicator and fix double counting

innerdvations committed:
Merge branch 'develop' into enhancement/transfer-chunk-progress

Bassel17 commented on packages/core/strapi/src/cli/utils/helpers.ts:
```diff
@@ -19,13 +19,40 @@ const readableBytes = (bytes: number, decimals = 1, padStart = 0) => {
+  }
+
+  const floored = floorToDecimal(value, decimals);
+  const result = `${floored.toFixed(decimals)}${unit}`;
```
I am a little confused do we actually need `.toFixed` here ?

innerdvations replied:
Yeah, this is what keeps it from flickering when we move from something like "29.9" to "30" and keeps it as "30.0"

Bassel17 reviewed (dismissed):
Overall looks good !
just left a small comment

innerdvations committed:
chore: clean up code

innerdvations committed:
Merge branch 'develop' into enhancement/transfer-chunk-progress

jhoward1994 reviewed (dismissed):
The code looks good to me. I think I'll need a bit of help to set up an asset transfer, let me know if you want me to test that too

innerdvations commented:
Some more QA revealed that this corrupts files; investigating it now

innerdvations committed:
Merge branch 'develop' into enhancement/transfer-chunk-progress

innerdvations committed:
Merge branch 'develop' into enhancement/transfer-chunk-progress

innerdvations committed:
fix: data transfer memory leak

innerdvations committed:
fix: yarn.lock

innerdvations committed:
chore: fix memory logging

innerdvations committed:
ci: complex project remote transfer

innerdvations committed:
Merge branch 'develop' into enhancement/transfer-chunk-progress

innerdvations committed:
enhancement: optimized transfer mode

innerdvations committed:
test(cli): data transfer and env passthrough

innerdvations committed:
chore: only send one message warning of legacy mode

innerdvations committed:
enhancement: show transfer totals and estimated time

innerdvations committed:
test(cli): fix deterministic transfer test files

innerdvations committed:
fix: push and pull shared utils backwards compatibility

innerdvations committed:
fix(data-transfer): extract legacy asset chunk parsing and tighten transfer logging/test coverage

innerdvations committed:
enhancement: checksum negotiation

innerdvations committed:
enhancement: show skipped file warnings on client

innerdvations committed:
fix: transfer diagnostics

innerdvations committed:
test: fix open handle

innerdvations committed:
fix: clear stall timeout for assets

innerdvations committed:
chore: fix misleading comments and variables

innerdvations committed:
test: fix misleading test

innerdvations committed:
test: fix typo

innerdvations committed:
test: make checks deterministic, less flaky

innerdvations committed:
enhancement(data-transfer): speed up asset totals; widen assets start reply window on remote pull

innerdvations committed:
Merge branch 'develop' into enhancement/transfer-chunk-progress

innerdvations committed:
Merge branch 'develop' into enhancement/transfer-chunk-progress

innerdvations committed:
fix(data-transfer): harden WebSocket JSON serialization for transfer frames

innerdvations committed:
fix(data-transfer): more transfer hardening

innerdvations committed:
test: fix test imports

innerdvations committed:
Merge branch 'develop' into enhancement/transfer-chunk-progress

innerdvations committed:
Merge branch 'develop' into enhancement/transfer-chunk-progress

innerdvations committed:
fix: await async write

innerdvations committed:
fix(data-transfer): resolve push transfer deadlock and harden async writes

- Extract createAssetsDestinationWritable so Writable callbacks run before
  uploadStream completes (same WS batch as PassThrough chunks).
- Add writable-async-write (write callback + drain/finished race; avoid hang
  on destroy).
- Wire push/pull, remote-source, file & directory sources to shared write().
- Fire-and-forget pull flush: Promise.resolve(flush).catch(onError); guard
  missing stream inside try.
- Add regression tests (assets writable, writable-async-write, handler checks).

innerdvations committed:
fix(data-transfer): write push stream batches sequentially

Use a for-loop with await write() instead of Promise.all over msg.data
so non-asset stages respect one in-flight write per objectMode Writable
and backpressure from writable-async-write.

- Validate minChunksForBackpressure in assertReadStreamBackpressure
- Add engine test for non-Buffer asset chunk byte progress (counts as 1)
- Assert push.ts keeps sequential msg.data handling in static handler test

innerdvations commented:
New experimental: `0.0.0-experimental.efc28eaabf4e75159e06914d9f2f8059073b806f`

cc @butcherZ

innerdvations committed:
fix(data-transfer): align push streamAsset with remote-source and harden tests

- Push handler: combine stream/end under one branch, error when start is missing
  or action is invalid; shorten stage write comments.
- Engine version-matching tests: use a fresh createDestination() per engine so
  parallel transfers do not share destination writables (MaxListeners warnings).
- File destination tests: mock createWriteStream with a new Writable per call.
- CLI transfer tests: mock progress.stream so transfer::finish runs after transfer
  and clears the progress setInterval (fixes Jest worker hang).
- Misc test cleanup: assets-destination timeout clearTimeout, collect listeners,
  writable-async-write teardown; tighten push/static test descriptions.

innerdvations committed:
test: remove parity test

innerdvations committed:
fix(data-transfer): harden collect() and stabilize transfer tests

- collect(): settle once, remove listeners on resolve/reject, avoid double completion
- engine tests: add expectHeapGrowthWithinNoise for heap smoke checks
- CLI transfer tests: console spies in beforeAll; jest.restoreAllMocks in afterAll
- stream test: remove removeAllListeners workaround

innerdvations commented:
`0.0.0-experimental.11f10e472f3e424651fbe8ee40a844786e8d836f`

butcherZ reviewed (approved):
QAed, and it works between two strapi instances locally, might still have issue pushing to remote, but let's release this fix first so we can take our time to explore the issue with cloud 😄
